File: sender/java/src/main/java/org/file/qrcode/protocol/SessionId.java
package org.file.qrcode.protocol;

/**
 * L5 会话标识（CONTRACT.md 第 7 节）。
 *
 * sessionId = SHA256(sha256(文件内容) ‖ T(2B 大端) ‖ K(3B 大端) ‖ codec(1B)) 的前 4 字节，
 * 按大端解释为 uint32。确定性推导使同一文件同一参数的重复传输天然续接。
 */
public final class SessionId {

    private SessionId() {
    }

    public static int derive(byte[] contentSha256, int T, int K, int codec) {
        byte[] buf = new byte[contentSha256.length + 6];
        System.arraycopy(contentSha256, 0, buf, 0, contentSha256.length);
        int p = contentSha256.length;
        buf[p] = (byte) ((T >>> 8) & 0xFF);
        buf[p + 1] = (byte) (T & 0xFF);
        buf[p + 2] = (byte) ((K >>> 16) & 0xFF);
        buf[p + 3] = (byte) ((K >>> 8) & 0xFF);
        buf[p + 4] = (byte) (K & 0xFF);
        buf[p + 5] = (byte) codec;
        byte[] d = StreamCodec.sha256(buf);
        return ((d[0] & 0xFF) << 24) | ((d[1] & 0xFF) << 16)
                | ((d[2] & 0xFF) << 8) | (d[3] & 0xFF);
    }

    /** 8 位小写 hex 表示 */
    public static String toHex(int sessionId) {
        return String.format("%08x", sessionId & 0xFFFFFFFFL);
    }
}

File: receiver/QrBinary/Protocol/SessionId.swift
//
//  SessionId.swift
//  QrBinary
//
//  L5 会话标识的确定性推导。对应 CONTRACT.md 第 7 节。
//  正常接收时 sessionId 直接读帧头，这里的推导仅用于完成后的自检与测试向量比对。
//

import Foundation

enum SessionId {

    /// sessionId = SHA256( sha256(文件内容) ‖ T(2B) ‖ K(3B) ‖ codec(1B) ) 的前 4 字节，大端
    static func derive(contentSha256: [UInt8], T: Int, K: Int, codec: Codec) -> UInt32 {
        var buf = contentSha256
        ByteOps.appendUInt16(&buf, UInt16(T))
        ByteOps.appendUInt24(&buf, UInt32(K))
        buf.append(codec.rawValue)
        let digest = StreamCodec.sha256(buf)
        return ByteOps.readUInt32(digest, 0)
    }

    /// 8 位小写十六进制，用于进度文件名与 UI 占位
    static func hex(_ sessionId: UInt32) -> String {
        String(format: "%08x", sessionId)
    }
}
